lib/OPCUA/Open62541/Test/Logger.pm

```diff
@@ -0,0 +1,160 @@
+use strict;
+use warnings;
+
+package OPCUA::Open62541::Test::Logger;
+use Carp;
+use POSIX;
+use Time::HiRes qw(gettimeofday time sleep);
+
+use Test::More;
+
+sub planning {
+    # number of ok(), pass() and fail() calls in this code
+    return 3;
+}
+
+sub new {
+    my $class = shift;
+    my $self = { @_ };
+    $self->{logger}
+	or croak "$class logger not given";
+
+    return bless($self, $class);
+}
+
+sub writelog {
+    my ($context, $level, $category, $message) = @_;
+    $context->printf("%d.%06d %d %s: %s\n",
+	gettimeofday(), $level, $category, $message);
+    $context->flush();
+}
+
+sub file {
+    my $self = shift;
+    my $file = shift;
+
+    ok(open(my $fh, '>', $file), "open log file")
+	or return fail "open '$file' for writing failed: $!";
+    $self->{logger}->setCallback(\&writelog, $fh, undef);
+    pass "set log callback";
+    $self->{file} = $file;
+}
+
+sub pid {
+    my $self = shift;
+    $self->{pid} = shift if @_;
+    return $self->{pid};
+}
+
+sub loggrep {
+    my $self = shift;
+    my ($regex, $timeout, $count) = @_;
+
+    my $end;
+    $end = time() + $timeout if $timeout;
+
+    do {
+	my $kid = waitpid($self->{pid}, WNOHANG);
+	if ($kid > 0 && $? != 0) {
+	    # child terminated with failure
+	    fail "no log grep match, child '$self->{pid}' failed: $?";
+	    return;
+	}
+	open(my $fh, '<', $self->{file}) or
+	    return fail "open '$self->{file}' for reading failed: $!";
+	my @match = grep { /$regex/ } <$fh>;
+	if (!$count && @match or $count && @match >= $count) {
+	    pass "log grep match";
+	    return wantarray ? @match : $match[0]
+	}
+	close($fh);
+	# pattern not found
+	if ($kid == 0) {
+	    # child still running, wait for log data
+	    sleep .1;
```

Comment:
sleep takes an integer. sleep .1 is equivalent to sleep 0, i.e. a noop. Please use Time::HiRes